### PR TITLE
Issue #9897: "final" modifier in pattern variable definition crashes Checkstyle

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #9897: "final" modifier in pattern variable definition crashes Checkstyle

Check regression reports:
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_checks-nonjavadoc-error/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_checks-only-javadoc-error/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_part1/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_part2/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_part3/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_part4/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_part5/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_part6/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_seventu-check-regression_part_1/index.html
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/diff_seventu-check-regression_part_2/index.html

AST regression report:
https://nmancus1.github.io/issue-9897_check_diff_reports_2021_04_19/antlr-diff/index.html

### CODE REVIEW

**Comment by pbludov** [comment]
@nmancus1 I extracted javac16 config to a separate issue #9907

**Review by romani — APPROVED** [review]
ok to merge.

**Comment by romani** [comment]
@pbludov , as we have problems with jdk16 builds in your PR, we should not block this PR by this.
Lets finish review of this PR and merge it sooner.

**Review by strkkk — APPROVED** [review]

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt [hunk 1]
```diff
@@ -1269,6 +1269,7 @@ CLASS_DEF -> CLASS_DEF [13:0]
     |       |   |           |       |   |       |--LITERAL_INSTANCEOF -> instanceof [168:25]
     |       |   |           |       |   |       |   |--IDENT -> o1 [168:22]
     |       |   |           |       |   |       |   `--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [168:36]
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt:1271]: Isn't it wrong that this is in a java 14 file?

**Review by rnveach — APPROVED** [review]

**Comment by rnveach** [comment]
Is this blocked by https://github.com/checkstyle/checkstyle/pull/9908 or will that other be made redundant?

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt (hunk 1 above)
Comment (nmancus1) [reply src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt:1271]: Syntactically, yes, but having an empty/ dummy `MODIFIERS` ast here isn't really hurting anything. We use the `MODIFIERS` ast in this way for all constructs (it is always created, even if no modifiers exist). Pattern variables were still in preview until Java 16, so I don't think that it is a big deal that we just stick with final implementation.

**Comment by nmancus1** [comment]
> Is this blocked by #9908 or will that other be made redundant?

See https://github.com/checkstyle/checkstyle/pull/9902#issuecomment-825662057

**Review by pbludov — APPROVED** [review]

**Pull request merged**